README.md
# Crypto Hash

The SHA-1 hash function implemented from scratch using pure dart.

I might add support for SHA-256 in the future.

## Installing

For now, Crypto Hash is not available through pub dev, but you can still install it by adding the following to your dependencies:

```yaml
dependencies:
  crypto_hash:
    git:
      url: https://github.com/lucas979797/dart-crypto-hash
      ref: main
```

## Usage

A simple usage example:

```dart
import 'dart:convert';

import 'package:crypto_hash/crypto_hash.dart';

void main() {
  
  var hash = ShaOne.hashString('Hello World!');
  print(hash); // prints 2ef7bde68ce5404e97d5f042f95f89f1c232871

  // Unicode:
  var testString2 = 'こんにちは世界！';
  var hash2 = ShaOne.hashBytes(utf8.encode(testString2));
  print(hash2); // prints f66b68c765fae864106c7352d06be5df5ab510c3

  // Hashing a file: (see example directory for lorem.txt)
  var hash3 = ShaOne.hashFile('example/lorem.txt');
  print(hash3); // prints 4b0b2b74dab6099447a0471d6b390ce21055d508
}

```

## Features and bugs

Let me know if you find any...

## Other

I built this library to improve my understanding of hashing, and it is my first time implementing a (semi) secure hashing algorithm.

I found the following resources useful for learning:

- [Lecture 21: SHA-1 Hash Function by Christof Paar](https://www.youtube.com/watch?v=JIhZWgJA-9o)
- [How Hash Algorithms Work](https://www.metamorphosite.com/one-way-hash-encryption-sha1-data-software)
- [The Wikipedia page](https://en.wikipedia.org/wiki/SHA-1)

I enjoyed doing the project in Dart, but if I did it again i'd just do it in C because there is a fair bit of extra manipulation that has to be done in Dart due to the language only supporting 64 bit integers.


CHANGELOG.md
# Changelog

## 0.4.0

- Added convenience method to `ShaOne` to hash a string
- Changed `ShaOne.hashData` to `ShaOne.hashBytes`

## 0.3.0

- Added convenience methods to `ShaOne` to hash a file from its path
- Added tests

## 0.2.0

- Changed API to use a static reference to `ShaOne`

## 0.1.0

- Initial version


test/crypto_hash_test.dart
import 'package:crypto_hash/crypto_hash.dart';
import 'package:test/test.dart';

void main() {
  group('Group 1', () {
    setUp(() {});

    test('Simple test', () {
      var testStr = 'A Test';
      var hash = ShaOne.hashBytes(testStr.codeUnits);
      expect(hash, '8f0c0855915633e4a7de19468b3874c8901df043');
    });

    test('String api test', () {
      var testStr = 'Hello World!';
      var hash = ShaOne.hashString(testStr);
      expect(hash, '2ef7bde68ce5404e97d5f042f95f89f1c232871');
    });

    test('Empty test', () {
      var testStr = '';
      var hash = ShaOne.hashBytes(testStr.codeUnits);
      expect(hash, 'da39a3ee5e6b4b0d3255bfef95601890afd80709');
    });

    test('448 test', () {
      var testStr = 'abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq';
      var hash = ShaOne.hashBytes(testStr.codeUnits);
      expect(hash, '84983e441c3bd26ebaae4aa1f95129e5e54670f1');
    });

    test('896 test', () {
      var testStr =
          'abcdefghbcdefghicdefghijdefghijkefghijklfghijklmghijklmnhijklmnoijklmnopjklmnopqklmnopqrlmnopqrsmnopqrstnopqrstu';
      var hash = ShaOne.hashBytes(testStr.codeUnits);
      expect(hash, 'a49b2446a02c645bf419f995b67091253a04a259');
    });

    test('1 mil a test', () {
      var testStr = 'a' * 1000000;
      var hash = ShaOne.hashBytes(testStr.codeUnits);
      expect(hash, '34aa973cd4c4daa4f61eeb2bdbad27316534016f');
    });

    test('If poem test', () {
      var hash = ShaOne.hashFile('test/if.txt');
      expect(hash, '9d3cd5d6e818f277d7b63b86d843410fe6bbda75');
    });

    test('Japanese test', () async {
      var hash = await ShaOne.hashFileAsync('test/boukendeshodesho.txt');
      expect(hash, 'b23ea7fae3b74b97efcda06095c392b09792de2b');
    });

    // this test takes about 20 seconds to compute the hash as the string is 1GB in length
    test('Final boss test', () {
      var testStr =
          'abcdefghbcdefghicdefghijdefghijkefghijklfghijklmghijklmnhijklmno' *
              16777216;
      var hash = ShaOne.hashBytes(testStr.codeUnits);
      expect(hash, '7789f0c9ef7bfc40d93311143dfbe69e2017f592');
    });
  });
}


lib/src/sha_one.dart
import 'dart:convert';
import 'dart:io';

import 'package:crypto_hash/src/utils.dart';

import 'constants.dart';
import 'context.dart';
import 'data.dart';

// Contains the hashing algorithm business logic
/// Contains methods that hash data using the SHA-1 hashing algorithm.
class ShaOne {
  static const _K1 = 0x5A827999;
  static const _K2 = 0x6ED9EBA1;
  static const _K3 = 0x8F1BBCDC;
  static const _K4 = 0xCA62C1D6;
  static const _STAGE_COUNT = 80;

  /// Hashes [data] using the SHA-1 hashing algorithm.
  /// [data] must effectively be 8 bit integers.
  /// Bits past the first byte in each int will be discarded.
  static String hashBytes(List<int> data) {
    var context = Sha1Context();
    var dataToHash = DataToHash(data);
    while (dataToHash.hasMoreSchedules) {
      var schedule = dataToHash.getNextSchedule();
      _processSchedule(context, schedule);
    }
    return context.getHash();
  }

  /// Hashes a String [s] by encoding it with utf8 and hashing the bytes. Accepts any Unicode string.
  static String hashString(String s) {
    return hashBytes(utf8.encode(s));
  }

  /// Hashes the file located at [path]. Reads the file synchronously.
  static String hashFile(String path) {
    return hashBytes(File(path).readAsBytesSync());
  }

  /// Hashes the file located at [path]. Reads the file asynchronously.
  static Future<String> hashFileAsync(String path) async {
    return hashBytes(await File(path).readAsBytes());
  }

  // The compression function
  static void _processSchedule(Sha1Context ctx, List<int> schedule) {
    var a = ctx.a, b = ctx.b, c = ctx.c, d = ctx.d, e = ctx.e;
    for (var j = 0; j < _STAGE_COUNT; j++) {
      var fRes = 0;
      if (j < 20) {
        // d should have 0s as 32 msbs so ~b's 1's in the 32 msbs should be cut off
        // brackets around the bitwise sections are absolutely necessary
        fRes += _K1 + ((b & c) | ((~b) & d));
      } else if (j < 40) {
        fRes += _K2 + (b ^ c ^ d);
      } else if (j < 60) {
        fRes += _K3 + ((b & c) | (b & d) | (c & d));
      } else {
        fRes += _K4 + (b ^ c ^ d);
      }

      // add together and throw away overflow bit
      var temp = (fRes + e + leftRotate32(a, 5) + schedule[j]) & BIT_MASK_32;
      // swap values
      e = d;
      d = c;
      c = leftRotate32(b, 30);
      b = a;
      a = temp;
      // For dev, check there is never any overflow
      assert(a == a & BIT_MASK_32);
      assert(b == b & BIT_MASK_32);
      assert(c == c & BIT_MASK_32);
      assert(d == d & BIT_MASK_32);
      assert(e == e & BIT_MASK_32);
    }
    ctx.addAndTruncate(a, b, c, d, e);
  }

  ShaOne._();
}


lib/src/data.dart
import 'dart:typed_data';

import 'package:crypto_hash/src/constants.dart';
import 'package:crypto_hash/src/utils.dart';
import 'package:typed_data/typed_buffers.dart';

// The data caretaker.
/// Serves up the next block of data for the SHA-1 hasing algorithm.
/// Signals when processing is done with [hasMoreSchedules].
/// Adds padding to final block.
class DataToHash {
  // the block size in bytes (64)
  static const BLOCK_SIZE = 512 ~/ 8;
  // the number of 32 bit words required for the schedule
  static const WORD32_COUNT = 80;
  // 9 bytes for byte divider plus 64 bit original message length byte count
  static const DIVIDER_AND_SIZE = 9;

  // longer messages are not supported by both Dart VM and JS
  // 2^53 - 1
  static const MAX_BYTE_LEN = 0x0003ffffffffffff;

  // Allows retreival of uint32 from a list of uint8
  late ByteData data;
  int nextByte = 0;
  late int lastByteOffset;

  DataToHash(List<int> data) {
    // initialise the random access buffer
    var uint8Buf = Uint8Buffer();
    uint8Buf.addAll(data);
    _addPadding(uint8Buf);
    lastByteOffset = uint8Buf.lengthInBytes - 1;
    this.data = uint8Buf.buffer.asByteData();
  }

  void _addPadding(Uint8Buffer uint8buf) {
    var originalLen = uint8buf.lengthInBytes;

    // separator byte
    uint8buf.add(0x80);
    // 512 bits = 64 bytes
    var mod = (uint8buf.lengthInBytes * 8) % 512;
    // 448 bits = 56 bytes
    if (mod < 448) {
      // if was less then 512 - mod is number of bytes needed to make it 448
      uint8buf.addAll(List.generate((448 - mod) ~/ 8, (i) => 0x0));
    } else if (mod > 448) {
      uint8buf.addAll(List.generate(((512 - mod) ~/ 8) + 56, (i) => 0x0));
    }
    // else there was the perfect number of bits to just add the length.
    var buffLen = uint8buf.lengthInBytes;

    assert(buffLen % 64 == 56);

    uint8buf.addAll(generateLengthBytes(originalLen * 8));

    if (uint8buf.lengthInBytes > MAX_BYTE_LEN) {
      throw UnsupportedError('Cannot hash data of more than 2^53 bytes');
    }
  }

  List<int> generateLengthBytes(int buffLen) {
    var lenBytes = <int>[];
    for (var i = 56; i >= 0; i -= 8) {
      lenBytes.add(buffLen >> i);
    }
    return lenBytes;
  }

  /// Generates the message schedule for an iteration.
  /// Schedule consists of 80 32 bit words derived from a 512 bit block
  List<int> getNextSchedule() {
    var schedule = List.filled(80, 0, growable: false);
    for (var i = 0; i < WORD32_COUNT; i++) {
      if (i < 16) {
        schedule[i] = data.getUint32(nextByte);
        nextByte += 4;
      } else {
        // xor to derive the rest of the schedule
        var xored = schedule[i - 3] ^
            schedule[i - 8] ^
            schedule[i - 14] ^
            schedule[i - 16];
        schedule[i] = leftRotate32(xored, 1);
      }
      assert(schedule[i] == schedule[i] & BIT_MASK_32);
    }

    return schedule;
  }

  bool get hasMoreSchedules => nextByte < lastByteOffset;
}
